consensus/src/liveness/proposal_status_tracker.rs

```diff
@@ -0,0 +1,146 @@
+    }
+
+    fn last_consecutive_statuses_matching<F>(&self, matcher: F) -> usize
+    where
+        F: Fn(&NewRoundReason) -> bool,
+    {
+        let mut last_consecutive_matching_count = 0;
+        for reason in self.past_round_statuses.iter().rev() {
+            if matcher(reason) {
+                last_consecutive_matching_count += 1;
+            } else {
+                break;
+            }
+        }
+        last_consecutive_matching_count
+    }
+
+    fn compute_failure_window(&mut self) {
+        self.last_consecutive_success_count =
+            self.last_consecutive_statuses_matching(|reason| match reason {
+                NewRoundReason::QCReady => true,
+                NewRoundReason::Timeout(timeout_reason) => match timeout_reason {
+                    RoundTimeoutReason::Unknown => true,
+                    RoundTimeoutReason::ProposalNotReceived => true,
+                    RoundTimeoutReason::PayloadUnavailable { .. } => false,
+                    RoundTimeoutReason::NoQC => true,
+                },
+            });
+        if self.last_consecutive_success_count == 0 {
+            self.window *= 2;
+            self.window = self.window.max(self.max_window);
+        } else if self.last_consecutive_success_count == self.past_round_statuses.len() {
+            self.window = 2;
+        }
+    }
+
+    fn get_exclude_authors(&self) -> HashSet<Author> {
+        let mut exclude_authors = HashSet::new();
+
+        let limit = self.window;
+        for round_reason in self.past_round_statuses.iter().rev().take(limit) {
+            if let NewRoundReason::Timeout(RoundTimeoutReason::PayloadUnavailable {
+                missing_authors,
+            }) = round_reason
+            {
+                for author_idx in missing_authors.iter_ones() {
+                    if let Some(author) = self.ordered_authors.get(author_idx) {
+                        exclude_authors.insert(*author);
+                    }
+                }
+            }
+        }
+
+        exclude_authors
+    }
+}
+
+impl TPastProposalStatusTracker for Mutex<ExponentialWindowFailureTracker> {
+    fn push(&self, status: NewRoundReason) {
+        self.lock().push(status)
+    }
+}
+
+pub struct OptQSPullParamsProvider {
+    enable_opt_qs: bool,
+    failure_tracker: Arc<Mutex<ExponentialWindowFailureTracker>>,
+}
+
+impl OptQSPullParamsProvider {
+    pub fn new(
+        enable_opt_qs: bool,
+        failure_tracker: Arc<Mutex<ExponentialWindowFailureTracker>>,
+    ) -> Self {
+        Self {
+            enable_opt_qs,
+            failure_tracker,
+        }
+    }
+}
+
+impl TOptQSPullParamsProvider for OptQSPullParamsProvider {
+    fn get_params(&self) -> Option<OptQSPayloadPullParams> {
+        if !self.enable_opt_qs {
+            return None;
+        }
+
+        let tracker = self.failure_tracker.lock();
+
+        let opt_batch_txns_pct = if tracker.last_consecutive_success_count < tracker.window {
```

Comment:
What is the rationale for either 0 or 50 percent?

Reply:
0 is straight-forward basically no opt_batch for the block. 50 is to restrict 50% block space to proofs. If we have a spike in mempool txns, QS is going to create a lot of proofs than consensus can pull for a block. This will lead to a situation where consensus doesn't keep up with the proofs so it never pulls opt batches. The downside is we might be expiring some proofs but this will allow us to reach an equlibirum state and pack opt batches.